File: src/glottisdale/collage/phonotactics.py
"""Phonotactic scoring for natural-sounding syllable ordering."""

import random

from glottisdale.types import Syllable

# Sonority scale for ARPABET phonemes (higher = more sonorous)
_SONORITY = {}

# 1: Stops
for p in ("P", "B", "T", "D", "K", "G"):
    _SONORITY[p] = 1

# 2: Affricates
for p in ("CH", "JH"):
    _SONORITY[p] = 2

# 3: Fricatives
for p in ("F", "V", "TH", "DH", "S", "Z", "SH", "ZH", "HH"):
    _SONORITY[p] = 3

# 4: Nasals
for p in ("M", "N", "NG"):
    _SONORITY[p] = 4

# 5: Liquids
for p in ("L", "R"):
    _SONORITY[p] = 5

# 6: Glides
for p in ("W", "Y"):
    _SONORITY[p] = 6

# Illegal English onsets (these sounds cannot start a word/syllable)
_ILLEGAL_ONSETS = {"NG", "ZH"}

# IPA sonority mapping for BFA phonemes
_IPA_VOWELS = set("aeiouɪɛæɑɒɔʊəɜɐʌ")
_IPA_STOPS = set("pbtdkgʔ")
_IPA_NASALS = set("mnɲŋɴ")
_IPA_FRICATIVES = set("fvθðszʃʒçxɣhɦ")
_IPA_LATERALS = set("lɫɬɮ")
_IPA_RHOTICS = {"r", "ɹ", "ɾ", "ɽ", "ʁ", "ʀ"}
_IPA_GLIDES = {"j", "w", "ɥ"}
_IPA_DIPHTHONG_STARTS = {"aɪ", "aʊ", "eɪ", "oʊ", "ɔɪ"}

# IPA illegal onsets (velar nasal can't start English syllables)
_IPA_ILLEGAL_ONSETS = {"ŋ"}


def _is_ipa(label: str) -> bool:
    """Heuristic: IPA labels use lowercase/Unicode, ARPABET uses uppercase ASCII."""
    if not label:
        return False
    return label[0].islower() or not label[0].isascii()


def _ipa_sonority(label: str) -> int:
    """Return sonority value for an IPA phoneme label."""
    if not label:
        return 0
    if any(label.startswith(d) for d in _IPA_DIPHTHONG_STARTS):
        return 7
    if label[0] in _IPA_VOWELS or label.rstrip("ːˑ") in _IPA_VOWELS:
        return 7
    if label in _IPA_GLIDES or label[0] in {"j", "w", "ɥ"}:
        return 6
    if label in _IPA_RHOTICS or label[0] in {"ɹ", "ɾ", "r"}:
        return 5
    if label[0] in _IPA_LATERALS:
        return 5
    if label[0] in _IPA_NASALS:
        return 4
    if label[0] in _IPA_FRICATIVES:
        return 3
    if label[0] in _IPA_STOPS:
        return 1
    return 0


def sonority(label: str) -> int:
    """Return sonority value for a phoneme label (ARPABET or IPA).

    Strips stress digits for ARPABET (e.g. 'AH0' -> vowel). Returns 0 for unknown.
    """
    if _is_ipa(label):
        return _ipa_sonority(label)

    # ARPABET path
    # Strip trailing stress digits for vowel lookup
    base = label.rstrip("012")
    if base in _SONORITY:
        return _SONORITY[base]
    # Check if it's a vowel (anything with a stress digit, or known vowel base)
    if label != base or base in (
        "AA", "AE", "AH", "AO", "AW", "AY",
        "EH", "ER", "EY", "IH", "IY",
        "OW", "OY", "UH", "UW",
    ):
        return 7
    return 0


def score_junction(syl_a: Syllable, syl_b: Syllable) -> int:
    """Score the phonotactic quality of the junction between two syllables.

    Higher scores = more natural-sounding transitions.
    """
    if not syl_a.phonemes or not syl_b.phonemes:
        return 0

    last_phone = syl_a.phonemes[-1].label
    first_phone = syl_b.phonemes[0].label

    score = 0

    # Illegal onset penalty
    base_first = first_phone.rstrip("012")
    if base_first in _ILLEGAL_ONSETS or first_phone in _IPA_ILLEGAL_ONSETS:
        score -= 2

    # Hiatus penalty (vowel-vowel boundary)
    if sonority(last_phone) == 7 and sonority(first_phone) == 7:
        score -= 1

    # Sonority contour: coda should fall, onset should rise toward nucleus
    boundary_sonority = sonority(last_phone) + sonority(first_phone)
    if boundary_sonority <= 8:  # Both consonantal
        score += 1
    elif boundary_sonority >= 12:  # Both very sonorous
        score -= 1

    return score


def order_syllables(
    syllables: list[Syllable],
    seed: int | None = None,
    attempts: int = 5,
) -> list[Syllable]:
    """Reorder syllables to maximize phonotactic junction quality.

    Tries `attempts` random permutations and returns the best-scoring one.
    """
    if len(syllables) <= 1:
        return list(syllables)

    rng = random.Random(seed)

    def total_score(ordering: list[Syllable]) -> int:
        return sum(
            score_junction(ordering[i], ordering[i + 1])
            for i in range(len(ordering) - 1)
        )

    best = list(syllables)
    best_score = total_score(best)

    for _ in range(attempts):
        candidate = list(syllables)
        rng.shuffle(candidate)
        s = total_score(candidate)
        if s > best_score:
            best = candidate
            best_score = s

    return best

File: crates/core/src/language/phonotactics.rs
//! Phonotactic scoring for natural-sounding syllable ordering.

use rand::seq::SliceRandom;
use rand::SeedableRng;
use rand::rngs::StdRng;
use std::collections::HashSet;

use crate::types::Syllable;

lazy_static::lazy_static! {
    /// ARPABET consonant sonority scale.
    static ref ARPABET_SONORITY: std::collections::HashMap<&'static str, i32> = {
        let mut m = std::collections::HashMap::new();
        // 1: Stops
        for p in &["P", "B", "T", "D", "K", "G"] { m.insert(*p, 1); }
        // 2: Affricates
        for p in &["CH", "JH"] { m.insert(*p, 2); }
        // 3: Fricatives
        for p in &["F", "V", "TH", "DH", "S", "Z", "SH", "ZH", "HH"] { m.insert(*p, 3); }
        // 4: Nasals
        for p in &["M", "N", "NG"] { m.insert(*p, 4); }
        // 5: Liquids
        for p in &["L", "R"] { m.insert(*p, 5); }
        // 6: Glides
        for p in &["W", "Y"] { m.insert(*p, 6); }
        m
    };

    /// ARPABET illegal onsets.
    static ref ILLEGAL_ONSETS: HashSet<&'static str> = {
        ["NG", "ZH"].into_iter().collect()
    };

    /// Known ARPABET vowel bases.
    static ref ARPABET_VOWELS: HashSet<&'static str> = {
        [
            "AA", "AE", "AH", "AO", "AW", "AY",
            "EH", "ER", "EY", "IH", "IY",
            "OW", "OY", "UH", "UW",
        ].into_iter().collect()
    };

    /// IPA vowel characters.
    static ref IPA_VOWELS: HashSet<char> = {
        "aeiouɪɛæɑɒɔʊəɜɐʌ".chars().collect()
    };

    /// IPA stops.
    static ref IPA_STOPS: HashSet<char> = {
        "pbtdkgʔ".chars().collect()
    };

    /// IPA nasals.
    static ref IPA_NASALS: HashSet<char> = {
        "mnɲŋɴ".chars().collect()
    };

    /// IPA fricatives.
    static ref IPA_FRICATIVES: HashSet<char> = {
        "fvθðszʃʒçxɣhɦ".chars().collect()
    };

    /// IPA laterals.
    static ref IPA_LATERALS: HashSet<char> = {
        "lɫɬɮ".chars().collect()
    };

    /// IPA rhotics.
    static ref IPA_RHOTICS: HashSet<&'static str> = {
        ["r", "ɹ", "ɾ", "ɽ", "ʁ", "ʀ"].into_iter().collect()
    };

    /// IPA glides.
    static ref IPA_GLIDES: HashSet<&'static str> = {
        ["j", "w", "ɥ"].into_iter().collect()
    };

    /// IPA diphthong starts.
    static ref IPA_DIPHTHONG_STARTS: Vec<&'static str> = {
        vec!["aɪ", "aʊ", "eɪ", "oʊ", "ɔɪ"]
    };

    /// IPA illegal onsets.
    static ref IPA_ILLEGAL_ONSETS: HashSet<&'static str> = {
        ["ŋ"].into_iter().collect()
    };
}

fn is_ipa(label: &str) -> bool {
    if label.is_empty() {
        return false;
    }
    let first = label.chars().next().unwrap();
    first.is_lowercase() || !first.is_ascii()
}

fn ipa_sonority(label: &str) -> i32 {
    if label.is_empty() {
        return 0;
    }
    if IPA_DIPHTHONG_STARTS.iter().any(|d| label.starts_with(d)) {
        return 7;
    }
    let first = label.chars().next().unwrap();
    let stripped = label.trim_end_matches(['ː', 'ˑ']);
    if IPA_VOWELS.contains(&first) || (stripped.len() == 1 && IPA_VOWELS.contains(&stripped.chars().next().unwrap_or(' '))) {
        return 7;
    }
    if IPA_GLIDES.contains(label) || matches!(first, 'j' | 'w' | 'ɥ') {
        return 6;
    }
    if IPA_RHOTICS.contains(label) || matches!(first, 'ɹ' | 'ɾ' | 'r') {
        return 5;
    }
    if IPA_LATERALS.contains(&first) {
        return 5;
    }
    if IPA_NASALS.contains(&first) {
        return 4;
    }
    if IPA_FRICATIVES.contains(&first) {
        return 3;
    }
    if IPA_STOPS.contains(&first) {
        return 1;
    }
    0
}

/// Return sonority value for a phoneme label (ARPABET or IPA).
///
/// Strips stress digits for ARPABET. Returns 0 for unknown.
pub fn sonority(label: &str) -> i32 {
    if is_ipa(label) {
        return ipa_sonority(label);
    }

    // ARPABET path
    let base = label.trim_end_matches(|c: char| c.is_ascii_digit());
    if let Some(&son) = ARPABET_SONORITY.get(base) {
        return son;
    }
    // Check if it's a vowel
    if base != label || ARPABET_VOWELS.contains(base) {
        return 7;
    }
    0
}

/// Score the phonotactic quality of the junction between two syllables.
///
/// Higher scores = more natural-sounding transitions.
pub fn score_junction(syl_a: &Syllable, syl_b: &Syllable) -> i32 {
    if syl_a.phonemes.is_empty() || syl_b.phonemes.is_empty() {
        return 0;
    }

    let last_phone = &syl_a.phonemes.last().unwrap().label;
    let first_phone = &syl_b.phonemes[0].label;

    let mut score = 0;

    // Illegal onset penalty
    let base_first = first_phone.trim_end_matches(|c: char| c.is_ascii_digit());
    if ILLEGAL_ONSETS.contains(base_first) || IPA_ILLEGAL_ONSETS.contains(first_phone.as_str()) {
        score -= 2;
    }

    // Hiatus penalty (vowel-vowel boundary)
    if sonority(last_phone) == 7 && sonority(first_phone) == 7 {
        score -= 1;
    }

    // Sonority contour
    let boundary_sonority = sonority(last_phone) + sonority(first_phone);
    if boundary_sonority <= 8 {
        score += 1;
    } else if boundary_sonority >= 12 {
        score -= 1;
    }

    score
}

/// Reorder syllables to maximize phonotactic junction quality.
///
/// Tries `attempts` random permutations and returns the best-scoring one.
pub fn order_syllables(
    syllables: &[Syllable],
    seed: Option<u64>,
    attempts: usize,
) -> Vec<Syllable> {
    if syllables.len() <= 1 {
        return syllables.to_vec();
    }

    let mut rng = match seed {
        Some(s) => StdRng::seed_from_u64(s),
        None => StdRng::from_entropy(),
    };

    let total_score = |ordering: &[Syllable]| -> i32 {
        ordering
            .windows(2)
            .map(|w| score_junction(&w[0], &w[1]))
            .sum()
    };

    let mut best = syllables.to_vec();
    let mut best_score = total_score(&best);

    for _ in 0..attempts {
        let mut candidate = syllables.to_vec();
        candidate.shuffle(&mut rng);
        let s = total_score(&candidate);
        if s > best_score {
            best = candidate;
            best_score = s;
        }
    }

    best
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Phoneme;

    fn make_syl(phoneme_labels: &[&str]) -> Syllable {
        Syllable {
            phonemes: phoneme_labels
                .iter()
                .map(|l| Phoneme {
                    label: l.to_string(),
                    start: 0.0,
                    end: 0.1,
                })
                .collect(),
            start: 0.0,
            end: 0.1,
            word: "test".to_string(),
            word_index: 0,
        }
    }

    #[test]
    fn test_sonority_arpabet() {
        assert_eq!(sonority("AE1"), 7); // vowel
        assert_eq!(sonority("K"), 1); // stop
        assert_eq!(sonority("S"), 3); // fricative
        assert_eq!(sonority("N"), 4); // nasal
        assert_eq!(sonority("L"), 5); // liquid
        assert_eq!(sonority("W"), 6); // glide
    }

    #[test]
    fn test_sonority_ipa() {
        assert_eq!(sonority("a"), 7); // vowel
        assert_eq!(sonority("p"), 1); // stop
        assert_eq!(sonority("s"), 3); // fricative
        assert_eq!(sonority("n"), 4); // nasal
    }

    #[test]
    fn test_score_junction_consonant_consonant() {
        let a = make_syl(&["K"]);
        let b = make_syl(&["T"]);
        let score = score_junction(&a, &b);
        assert!(score >= 0); // consonant-consonant is ok
    }

    #[test]
    fn test_score_junction_vowel_vowel() {
        let a = make_syl(&["AE1"]);
        let b = make_syl(&["IY1"]);
        let score = score_junction(&a, &b);
        assert!(score < 0); // hiatus penalty
    }

    #[test]
    fn test_score_junction_illegal_onset() {
        let a = make_syl(&["AE1"]);
        let b = make_syl(&["NG"]);
        let score = score_junction(&a, &b);
        assert!(score < 0); // NG can't start a syllable
    }

    #[test]
    fn test_order_syllables_deterministic() {
        let syls = vec![
            make_syl(&["K", "AE1"]),
            make_syl(&["T", "IY1"]),
            make_syl(&["S", "AH0"]),
        ];
        let a = order_syllables(&syls, Some(42), 10);
        let b = order_syllables(&syls, Some(42), 10);
        // Same seed should give same result
        assert_eq!(a.len(), b.len());
        for (sa, sb) in a.iter().zip(b.iter()) {
            assert_eq!(sa.phonemes.len(), sb.phonemes.len());
        }
    }

    #[test]
    fn test_order_syllables_single() {
        let syls = vec![make_syl(&["K", "AE1"])];
        let result = order_syllables(&syls, None, 10);
        assert_eq!(result.len(), 1);
    }

    #[test]
    fn test_order_syllables_empty() {
        let result = order_syllables(&[], None, 10);
        assert!(result.is_empty());
    }
}
